### PR TITLE
Refactor DataStream

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
See Changes below. This was part of https://github.com/dmlc/gluon-nlp/pull/218. I split up the PR to make it easier to review / discuss.

### Essentials ###
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage
- [X] Code is well-documented

### Changes ###
- [X] Remove DataStream.transform() and encourage the use of Python generator expressions instead
- [X] Remove SimpleDataStream (without .transform(), any Python iterable implements the DataStream interface. Consequently there is no need for SimpleDataStream. Let's keep the class hierarchy simple.)
- [X] Refactor PrefetchingStream
  - Remove option to iterate over a list of Streams. This allows keeping the implementation simple. A list of streams can be transformed into a single stream using `zip`.
  - Add option to use multi-processing (with a single worker process)
- [X] Introduce DatasetStream to clarify when we expect a Stream over Datasets vs over individual samples
- [X] Refactor CorpusStream to iterate over CorpusDataset instead of single sentences
  - An iterator over single sentences can still be obtained via `itertools.chain.from_iterable(corpus_stream)`. However, it can be useful to iterate over bigger chunks in the stream to speed up other steps in the data pipeline.
  - Remove the `sampler` argument. It is not needed anymore. (Note that it still kept for LanguageModelStream, which internally now uses `itertools.chain.from_iterable(corpus_stream)`)
- [X] Fix random stability of PrefetchingStream
  - Set the python, numpy and mxnet seeds in the started thread / process deterministically based on the python random number generator state in the calling thread.

## Comments ##
- We may want to re-consider introducing `DatasetStream`. The only difference to `DataStream` is the docstring. It may or may not be a good idea to complicate the class hierarchy for this.

### CODE REVIEW

**Comment by leezu** [comment]
With this change, we should also define exactly what we consider to be a stream (or a valid input to a `SimpleDataStream`). Do we consider iterable objects (has an `__iter__`method which returns an iterator; or defines `__getitem__`) or iterator objects (has a `next` (Python 2) or `__next__ `(Python 3) method.) as valid input? The main difference is that an if we call `iter(iterable)` multiple times, we will get independent iterators, whereas we can only iterate once over a single iterator.

In the context of GluonNLP DataStream this means, do we want to allow iterating over a DataStream multiple times? Or do we consider a DataStream to be a throwaway object that should only be used once? (It is cheap to create a new DataStream).

This PR proposed using generator expression `transformed = (transform(s) for s in stream)` instead of of `stream.transform()`. Note that the generator expression always returns a iterator. `stream.transform()` however always returns a iterable, meaning that if the underlying stream is an iterable it is possible to iterate multiple times over the transformed stream (and in case the underlying stream is an iterator, it is only possible to iterate once over the transformed stream).

Consequently, using generator expressions only makes sense if we only require the stream to be an iterator. If we decide to require the Stream to be an iterable I will revert the respective change and add back the `.transform()` interface.

(Independently of this question, in the previous implementation `PrefetchingStream(stream).transform()` wouldn't work.)

A small example to illustrate the difference:

```
In [1]: import gluonnlp as nlp
In [2]: iterator = [1,2,3]
In [3]: iterable = (i for i in range(3))
In [4]: stream1 = nlp.data.SimpleDataStream(iterator)
In [5]: stream2 = nlp.data.SimpleDataStream(iterable)
In [6]: list(stream1)
Out[6]: [1, 2, 3]
In [7]: list(stream1)
Out[7]: [1, 2, 3]
In [8]: list(stream2)
Out[8]: [0, 1, 2]
In [9]: list(stream2)
Out[9]: []
```

File: gluonnlp/data/stream.py [hunk 1]
```diff
@@ -382,20 +312,23 @@ class _LanguageModelBPTTStream(DataStream):
         The length of each of the samples for truncated back-propagation-through-time (TBPTT).
     batch_size : int
         The number of samples in each batch.
+     sampler : mx.gluon.data.Sampler
```
Comment (eric-haibin-lin) [inline gluonnlp/data/stream.py]: nit: indentation

File: gluonnlp/data/stream.py [hunk 2]
```diff
@@ -453,3 +388,117 @@ def _write(data, target, buffers, seq_len, i, length):
...
+    will be initialized randomly based on the next 32 bit integer in the
+    python, numpy and mxnet random generator of the caller respectively
+    (random.getrandbits(32), numpy.random.randint(0, 2**32),
+    int(mx.nd.random.uniform(0, 2**32).asscalar())).
+
+    Parameters
+    ----------
+    stream : DataStream
+        Source stream.
+    num_prefetch : int, default 1
```
Comment (eric-haibin-lin) [inline gluonnlp/data/stream.py]: is `num_prefetch ` > 1 also supported?

Comment (leezu) [reply gluonnlp/data/stream.py]: It should be. num_prefetch sets the size of the queue. The Thread / Process tries to push as many elementsas possible to the queue.

**Comment by mli** [comment]
Job PR-235/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-235/15/index.html

File: gluonnlp/data/stream.py (hunk 2 above)
Comment (leezu) [reply gluonnlp/data/stream.py]: I updated the tests to run with different values of num_prefetch as well as thread / process.

File: gluonnlp/data/stream.py (hunk 1 above)
Comment (leezu) [reply gluonnlp/data/stream.py]: thanks! fixed

**Comment by mli** [comment]
Job PR-235/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-235/16/index.html

**Comment by mli** [comment]
Job PR-235/25 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-235/25/index.html

**Comment by mli** [comment]
Job PR-235/26 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-235/26/index.html

**Comment by mli** [comment]
Job PR-235/34 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-235/34/index.html

**Comment by mli** [comment]
Job PR-235/36 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-235/36/index.html

File: gluonnlp/data/stream.py [hunk 3]
```diff
@@ -453,3 +400,119 @@ def _write(data, target, buffers, seq_len, i, length):
             if has_token_buffered or self._last_batch == 'keep':
                 yield mx.nd.array(data).T, mx.nd.array(target).T, mx.nd.array(mask).T
         return
+
+
+class _Prefetcher(object):
+    """Interal shared prefetcher logic."""
```
Comment (eric-haibin-lin) [inline gluonnlp/data/stream.py]: nit: Interal -> Internal

File: gluonnlp/data/stream.py [hunk 4]
```diff
@@ -453,3 +400,119 @@ def _write(data, target, buffers, seq_len, i, length):
...
+
+    def next(self):
+        return self.__next__()
+
+    def __iter__(self):
+        return self
+
+
+class _ProcessPrefetcher(_Prefetcher, multiprocessing.Process):
+    """Interal multi-processing prefetcher."""
```
Comment (eric-haibin-lin) [inline gluonnlp/data/stream.py]: Internal

File: gluonnlp/data/stream.py [hunk 5]
```diff
@@ -453,3 +400,119 @@ def _write(data, target, buffers, seq_len, i, length):
...
+    def __init__(self, *args, **kwargs):
+        super(_ProcessPrefetcher, self).__init__(*args, **kwargs)
+        self.data_queue = multiprocessing.Queue(self.num_prefetch)
+        self.control_queue = multiprocessing.Queue()
+        self.daemon = True
+        self.start()
+
+
+class _ThreadPrefetcher(_Prefetcher, threading.Thread):
+    """Interal threaded prefetcher."""
```
Comment (eric-haibin-lin) [inline gluonnlp/data/stream.py]: Internal

File: gluonnlp/data/stream.py (hunk 2 above)
Comment (eric-haibin-lin) [reply gluonnlp/data/stream.py]: Oh okay I was under the wrong impression that this is the number of prefetching threads. nvm

**Comment by mli** [comment]
Job PR-235/38 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-235/38/index.html

**Pull request merged**